server/src/main/java/org/opensearch/action/search/SearchTransportService.java

```diff
@@ -337,15 +384,42 @@ private void sendExecuteFetch(
      */
     void sendExecuteMultiSearch(final MultiSearchRequest request, SearchTask task, final ActionListener<MultiSearchResponse> listener) {
         final Transport.Connection connection = transportService.getConnection(transportService.getLocalNode());
+
+        TransportRequestOptions options = getTransportRequestOptions(task, listener::onFailure, false);
+        if (options == null) {
+            return;
+        }
+
         transportService.sendChildRequest(
             connection,
             MultiSearchAction.NAME,
             request,
             task,
+            options,
             new ConnectionCountingHandler<>(listener, MultiSearchResponse::new, clientConnections, connection.getNode().getId())
         );
     }
 
+    static TransportRequestOptions getTransportRequestOptions(SearchTask task, Consumer<Exception> onFailure, boolean queryPhase) {
+        if (task != null && task.timeoutMills() > 0) {
+            long leftTimeMills;
+            if (queryPhase) {
+                // it's costly in query phase.
+                leftTimeMills = task.queryPhaseTimeout() - (System.currentTimeMillis() - task.startTimeMills());
```

Comment:
What's the motivation behind the `queryPhaseTimeoutPercentage` concept? I think it's going to depend on the query and the setup whether query or fetch phase takes longer and it doesn't seem super intuitive for a user to understand how to use this. For example a query that matches a lot of sparse documents using searchable snapshots might spend much longer in the fetch phase while a query that performs complex aggregations might spend a lot longer in the query phase.

Reply:
I hope to reserve some time for the subsequent phase as a backup measure, to ensure each stage can be allocated a certain amount of time. Of course, if the previous stage takes a very short time, it won't affect the remaining time available for the subsequent phases either.

If no such reservation is made, and a shard is blocked  in query phase and uses up all the time, even if it returns after the timeout, there won't be any executable time left for the subsequent stages, and the timeout would be meaningless in that case.